File: crates/compiler/frontend/typescript/src/host-capabilities.ts
// The host-fulfilled Capability ids the package being captured declares.
//
// A host capability is declared in `agent.toml` as `[[capabilities.host]]` and
// referred to as `Capability("host:<id>")`. The runtime never executes one; the
// embedding host answers the request (ADR-0025). The declarations live in the
// package manifest, which the confined interpreter never reads, so the trusted
// host bridge supplies them before the submitted module is evaluated and the
// minted Capability set becomes the builtin catalogue united with these.
//
// Nothing here is reachable from the public authoring entrypoint: a program
// that could declare its own host capabilities would be minting its own
// authority.

let declared: ReadonlySet<string> = new Set();

/** Supply the host capability ids the package manifest declares. */
export function setDeclaredHostCapabilities(ids: readonly string[]): void {
  declared = new Set(ids.map((id) => `host:${id}`));
}

/** Whether this package declares the host capability `reference` names. */
export function isDeclaredHostCapability(reference: string): boolean {
  return declared.has(reference);
}

/** The references this package mints, in a stable order, for a diagnostic. */
export function declaredHostCapabilities(): readonly string[] {
  return [...declared].sort();
}

File: crates/compiler/frontend/python/apxm_program/_host_capabilities.py
"""The host-fulfilled Capability ids the package being captured declares.

A host capability is declared in ``agent.toml`` as ``[[capabilities.host]]`` and
referred to as ``Capability("host:<id>")``. The runtime never executes one; the
embedding host answers the request (ADR-0025). The declarations live in the
package manifest, which the confined interpreter never reads, so the trusted
capture harness supplies them before the submitted module runs and the minted
Capability set becomes the builtin catalogue united with these.

This module is private. A program that could declare its own host capabilities
would be minting its own authority.
"""

from __future__ import annotations

from typing import Iterable

HOST_CAPABILITY_REF_PREFIX = "host:"

_declared: frozenset[str] = frozenset()


def set_declared_host_capabilities(ids: Iterable[str]) -> None:
    """Supply the host capability ids the package manifest declares."""
    global _declared
    _declared = frozenset(f"{HOST_CAPABILITY_REF_PREFIX}{identifier}" for identifier in ids)


def is_declared_host_capability(reference: str) -> bool:
    """Whether this package declares the host capability ``reference`` names."""
    return reference in _declared


def declared_host_capabilities() -> tuple[str, ...]:
    """The references this package mints, in a stable order, for a diagnostic."""
    return tuple(sorted(_declared))
